### PR TITLE
Add backslash escape support to containsSqlScriptDelimiters

### PR DESCRIPTION
**Pull request by geordiecoder**
Issue: SPR-17120

This change makes containsSqlScriptDelimiters keep track of backslash escapes while it determines if the current character is inside a literal.

N.B backslash escaped quotes are supported by MySQL / MariaDB but I don't believe they're supported by HSQLDB.

### CODE REVIEW

**Comment by pivotal-issuemaster** [comment]
@geordiecoder Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-projects/spring-framework&pullRequestId=1919)!

**Comment by sbrannen** [comment]
Thanks for the PR!

We're looking into it.

File: spring-jdbc/src/main/java/org/springframework/jdbc/datasource/init/ScriptUtils.java [hunk 1]
```diff
@@ -343,8 +343,22 @@ private static void appendSeparatorToScriptIfNecessary(StringBuilder scriptBuild
 	 */
 	public static boolean containsSqlScriptDelimiters(String script, String delim) {
 		boolean inLiteral = false;
+		boolean inEscape = false;
 		for (int i = 0; i < script.length(); i++) {
-			if (script.charAt(i) == '\'') {
+
+			char thisChar = script.charAt(i);
+			if (thisChar == '\\')
+			{
```
Comment (sbrannen) [inline spring-jdbc/src/main/java/org/springframework/jdbc/datasource/init/ScriptUtils.java]: Please move the opening curly bracket to the previous line as you see elsewhere in the codebase.

File: spring-jdbc/src/main/java/org/springframework/jdbc/datasource/init/ScriptUtils.java [hunk 2]
```diff
@@ -343,8 +343,22 @@ private static void appendSeparatorToScriptIfNecessary(StringBuilder scriptBuild
...
+				inEscape = !inEscape;
+				continue;
+			}
+			else if (inEscape)
+			{
```
Comment (sbrannen) [inline spring-jdbc/src/main/java/org/springframework/jdbc/datasource/init/ScriptUtils.java]: Please move the opening curly bracket to the previous line as you see elsewhere in the codebase.

File: spring-jdbc/src/main/java/org/springframework/jdbc/datasource/init/ScriptUtils.java [hunk 3]
```diff
@@ -343,8 +343,22 @@ private static void appendSeparatorToScriptIfNecessary(StringBuilder scriptBuild
 	 */
 	public static boolean containsSqlScriptDelimiters(String script, String delim) {
 		boolean inLiteral = false;
+		boolean inEscape = false;
 		for (int i = 0; i < script.length(); i++) {
-			if (script.charAt(i) == '\'') {
+
+			char thisChar = script.charAt(i);
+			if (thisChar == '\\')
```
Comment (sbrannen) [inline spring-jdbc/src/main/java/org/springframework/jdbc/datasource/init/ScriptUtils.java]: Please rename `thisChar` to `c` for consistency with `splitSqlScript()`.

**Review by sbrannen — CHANGES_REQUESTED** [review]

**Comment by geordiecoder** [comment]
Thanks for the feedback @sbrannen - I've pushed a commit to fix the code style issues.

**Review by sbrannen — APPROVED** [review]

**Comment by sbrannen** [comment]
This has been merged into `master` in 24ed6de6aaece130fa38e758712cedf6987212d0 and further refined in bdac39150f7137ab244590131d3d8b825ccc9e5d.

Thanks!

**Pull request closed without merging**